src/euler_one_qubit_decomposer.rs

```diff
@@ -10,15 +10,635 @@
+                let indices = slc.indices(len)?;
+                let mut out_vec: Vec<(String, Vec<f64>)> = Vec::new();
+                // Start and stop will always be positive the slice api converts
+                // negatives to the index for example:
+                // list(range(5))[-1:-3:-1]
+                // will return start=4, stop=2, and step=-1
+                let mut pos: isize = indices.start;
+                let mut cond = if indices.step < 0 {
+                    pos > indices.stop
+                } else {
+                    pos < indices.stop
+                };
+                while cond {
+                    if pos < len as isize {
+                        out_vec.push(self.gates[pos as usize].clone());
+                    }
+                    pos += indices.step;
+                    if indices.step < 0 {
+                        cond = pos > indices.stop;
+                    } else {
+                        cond = pos < indices.stop;
+                    }
+                }
+                Ok(out_vec.into_py(py))
+            }
+            SliceOrInt::Int(idx) => {
+                let len = self.gates.len() as isize;
+                if idx >= len || idx < -len {
+                    Err(PyIndexError::new_err(format!("Invalid index, {idx}")))
+                } else if idx < 0 {
+                    let len = self.gates.len();
+                    Ok(self.gates[len - idx.unsigned_abs()].clone().into_py(py))
+                } else {
+                    Ok(self.gates[idx as usize].clone().into_py(py))
+                }
+            }
+        }
+    }
+}
+
+fn circuit_kak(
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    k_gate: &str,
+    a_gate: &str,
+    simplify: bool,
+    atol: Option<f64>,
+) -> OneQubitGateSequence {
+    let mut lam = lam;
+    let mut theta = theta;
+    let mut phi = phi;
+    let mut circuit: Vec<(String, Vec<f64>)> = Vec::with_capacity(3);
+    let mut atol = match atol {
+        Some(atol) => atol,
+        None => DEFAULT_ATOL,
+    };
+    if !simplify {
+        atol = -1.0;
+    }
+    let mut global_phase = phase - (phi + lam) / 2.;
+    if theta.abs() < atol {
+        lam += phi;
+        // NOTE: The following normalization is safe, because the gphase correction below
+        // fixes a particular diagonal entry to 1, which prevents any potential phase
+        // slippage coming from _mod_2pi injecting multiples of 2pi.
+        lam = mod_2pi(lam);
+        if lam.abs() > atol {
+            circuit.push((String::from(k_gate), vec![lam]));
+            global_phase += lam / 2.;
+        }
+        return OneQubitGateSequence {
+            gates: circuit,
+            global_phase,
+        };
+    }
+    if (theta - PI).abs() < atol {
+        global_phase += phi;
+        lam -= phi;
+        phi = 0.;
+    }
+    if mod_2pi(lam + PI).abs() < atol || mod_2pi(phi + PI).abs() < atol {
```

Comment:
This hasn't transplanted the `allow_non_canonical` stuff from the Python side.  Do we need that?

Reply:
There was nothing in the user facing public api that I saw which used it, it was only defined on the private circuit constructor functions and nothing ever set it `False` so I opted to just remove the option.

Reply:
Sounds good to me.